[USB Host] HID driver component

roma-jam opened the pull request:
# Checklist

- [X] Component contains License
- [X] Component contains README.md
- [X] Component contains idf_component.yml file with `url` field defined
- [ ] Component was added to [upload job](https://github.com/espressif/idf-extra-components/blob/master/.github/workflows/upload_component.yml#L18)
- [ ] Component was added to [build job](https://github.com/espressif/idf-extra-components/blob/master/test_app/CMakeLists.txt#L8)
- [X] _Optional:_ Component contains unit tests
- [ ] CI passing

# Change description
USB Host HID driver component

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/idf-extra-components?pullRequest=175) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/idf-extra-components?pullRequest=175) before we can accept your contribution.<br/><hr/>**Roman Leonov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/idf-extra-components?pullRequest=175) it.</sub>

tore-espressif commented on usb/usb_host_hid/test/test_hid_basic.c:
```diff
@@ -0,0 +1,516 @@
+
+static void force_conn_state(bool connected, TickType_t delay_ticks)
+{
+    TEST_ASSERT_NOT_EQUAL(NULL, phy_hdl);
```
nitpick
```suggestion
    TEST_ASSERT_NOT_NULL(phy_hdl);
```

tore-espressif commented on usb/usb_host_hid/test/test_hid_basic.c:
```diff
@@ -0,0 +1,516 @@
+               dev_params->addr,
+               dev_params->iface_num);
+
+        data = hid_host_device_get_data(hid_device_handle,
```
Now I see where this function is used. We can discuss offline

tore-espressif commented on usb/usb_host_hid/test/test_hid_basic.c:
```diff
@@ -0,0 +1,516 @@
+               dev_params->iface_num);
+        TEST_ASSERT_EQUAL(ESP_OK, hid_host_device_close(hid_device_handle) );
+        break;
+    case HID_HOST_INTERFACE_EVENT_TRANSFER_ERROR:
```
This should not happen during the testing. Can we assert here?

tore-espressif commented on usb/usb_host_hid/test/test_hid_basic.c:
```diff
@@ -0,0 +1,516 @@
+    hid_host_dev_params_t *dev_params = hid_host_device_get_params(hid_device_handle);
+
+    TEST_ASSERT_NOT_NULL_MESSAGE(dev_params, "Device parameters not found");
+    TEST_ASSERT_EQUAL_MESSAGE(&user_arg_value, arg, "User argument has lost");
```
```suggestion
    TEST_ASSERT_PTR_EQUAL_MESSAGE(&user_arg_value, arg, "User argument has lost");
```

tore-espressif commented on usb/usb_host_hid/test/test_hid_basic.c:
```diff
@@ -0,0 +1,516 @@
+/*
```
Could we extend the tests? 
1. Test with multiple active interfaces (eg. keybour and mouse)
2. sudden disconnection test

roma-jam replied:
@tore-espressif  Yep, indeed. 

1. Just in case, I have all the devices with several active interfaces. How would you prefer this test to look like? 
2. Also, all disconnection for HID device is sudden. That was the one crucial idea, when I wrote the driver: HID device can appears or disappears any time. Or do you mean that disconnection should happen in the middle of the usb control transfer?

tore-espressif replied:
1. Test with a device that has both Mouse and Keyboard interfaces. Access them both concurently and make sure that there is no 'cross-talk' . This would probably required updates of the mock HID device, so we can leave it for later. It would be nice to have it now, if it is easy to implement, though
2. Yes, start a lot of transfers in one thread and call `force_conn_state(false, 1);` from another thread

roma-jam replied:
Sudden disconnection test marked with ignore for now.

roma-jam commented:
@tore-espressif, 
Thanks for such a full review! 
I will fix everything ASAP.

esp-saurabhbansal commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1465 @@
+ * @param[in] hidh_iface      Pointer to HID Interface configuration structure
+ * @return esp_err_t
+ */
+esp_err_t hid_class_request_report_descriptor(hid_iface_t *iface)
```
should this function be made static ?

esp-saurabhbansal commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1465 @@
+        .data = iface->report_desc
+    };
+
+    return usb_class_request_get_descriptor(iface->parent, &get_desc);
```
If the return value is not ESP_OK, I think memory that is allocated above in the following line will leak:
iface->report_desc = calloc(1, iface->report_desc_size);

tore-espressif replied:
Lifetime of iface->report_desc is equal to iface lifetime, right @roma-jam ? so we free this memory only after the device was detached/removed

roma-jam commented on usb/usb_host_hid/test/test_hid_basic.c:
```diff
@@ -0,0 +1,516 @@
+                                          &hid_test_task_handle));
+}
+
+TEST_CASE("memory_leakage", "[hid_host]")
+{
+    // Install USB and HID driver with the regular hid_host_test_callback
+    test_hid_setup(hid_host_test_callback);
+    // Tear down test
+    test_hid_teardown();
+    // Verify the memory leackage during test environment tearDown()
+}
```
Yes, I understood. 
I have made this test to make verification more easily. 
When there is a memory leakage, probably the first question would be: ok, where is a leakage? 
This test should distinguish two different scenarios: when the leakage in the hid driver install/remove  or when device is attached. 

But it is fair not to run this test with others, cause in that case there is not so much sense in it, that is true.

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+*/
+typedef struct {
+    bool create_background_task;            /**< When set to true, background task handling USB events is created.
+                                         Otherwise user has to periodically call hid_host_event_handler_task function */
```
Should be. 
Thanks for pointing out.

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+/**
+ * @brief USB HID Host open a device with specific device parameters
+ *
+ * @param[in] iface_handle     Pointer to an opened HID Device
```
Information has left from the previous version.
Thanks a lot for such a thorough review!

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+ *
+ * @param[in] hid_dev_handle    HID Device handle
+ * @param[in] report_id         ReportID
+ * @param[out] idle_rate        Idle rate
```
Sure, no problem.

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+/**
+ * @brief USB HID Host close device
+ *
+ * @param[in] hid_dev_handle   Device handle obtained from hid_host_open_device function
```
Yep, and not "a little".
Fixed.

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+ * @param[in] hid_dev_handle    HID Device handle
+ * @param[in] report_id         Report ID
+ * @param[out] report           Pointer to buffer for a report data
+ * @param[in/out] report_length Report data length
```
Thanks, also there will be an error, when the report buffer size is smaller that actual report data. 
Fixed.

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+ * @return esp_err_t
+ */
+esp_err_t hid_class_request_set_protocol(hid_host_device_handle_t hid_dev_handle,
+        uint8_t protocol);
```
Moreover, is should accept hid_report_protocol_t.

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+ *
+ * @return a uint8_t pointer to a data buffer
+ */
+uint8_t *hid_host_device_get_data(hid_host_device_handle_t hid_dev_handle,
```
Yes, that is fair.

roma-jam commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1465 @@
+    setup->wIndex = req->wIndex;
+    setup->wLength = req->wLength;
+
+    HID_RETURN_ON_ERROR( hid_control_transfer(hid_device,
```
Yeeees, good catch!

roma-jam commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1465 @@
+                        ESP_ERR_INVALID_STATE,
+                        "Unable to request report decriptor. Interface is not ready");
+
+    iface->report_desc = calloc(1, iface->report_desc_size);
```
I don't see any reason, why it is important here. calloc() should return a zero-filled memory.
But I can change that to malloc(), indeed.

tore-espressif replied:
Just to save few CPU cycles :) calloc calls memset which takes some time. But yeah, nothing critical

roma-jam commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1465 @@
+    usb_transfer_t *in_xfer;                /**< Pointer to IN transfer buffer */
+    hid_host_interface_event_cb_t user_cb;  /**< Interface application callback */
+    void *user_cb_arg;                      /**< Interface application callback arg */
+    hid_iface_state_t state;                /**< Interface ready for claiming */
```
Thanks for pointing out!

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,267 @@
+ * @param[in] hid_dev_handle  HID Device handle
+ * @return esp_err_t
+ */
+esp_err_t hid_host_device_start(hid_host_device_handle_t hid_dev_handle);
```
Yep, you are write. 
Moreover, I still questioning myself why we need two callbacks here, because one would be enough. 
Such as event loop usage. 

I guess, that I can leave that for a next version.

roma-jam commented on usb/usb_host_hid/test/test_hid_err_handling.c:
```diff
@@ -0,0 +1,136 @@
+static void test_install_hid_driver_with_wrong_config(void);
+static void test_claim_interface_without_driver(void);
+
+void (*test_hid_err_handling_case[TEST_HID_ERR_HANDLING_CASES])(void) = {
+    test_install_hid_driver_without_config,
+    test_install_hid_driver_with_wrong_config,
+    test_claim_interface_without_driver
+};
```
I just saw how you did the error_handling verification in cdc and tried to do that same (one unit test for every error). 
Do I really need to make a separate TEST_CASE for every error?

roma-jam replied:
Agreed, here they are not necessary. 
"overengineerd pointer to functions" - already removed them. :+1:

tore-espressif commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1465 @@
+        // reallocate the ctrl xfer buffer for new length
+        ESP_LOGI(TAG, "Change HID ctrl xfer size from %d to %d",
+                 hid_device->ctrl_xfer->data_buffer_size,
+                 (int) (dev_info.bMaxPacketSize0 + req->wLength));
```
Should it be:
```suggestion
                 (int) (USB_SETUP_PACKET_SIZE + req->wLength));
```
?

roma-jam commented:
@tore-espressif @esp-saurabhbansal 
Thanks you guys for the previous review. 
Sorry, that I had to update the code after, but I prepared a brief description what was done. 
 
A few changes were made: 
1. Solved all comments/nitpicks (through the code, remarks and so on)
2. Added test for verify uninstall HID driver while device is still present (helped to fix the bug with the wrong logical sequence during the uninstall: verify state of driver -> remove all available interfaces from the list)
3. Added second HID mock device with two independent interfaces (IFace1 and IFace2) to test connection of the USB Device with two HID Interfaces. Now the tests run on both HID mock devices (With one Interface* and with two Interfaces). 
4. Added a test for sudden disconnection of the HID device, while device stressed with the requests through the control EP. This test is ignored for now by the reason that this disconnection result in Core 0 panic'ed due to wrong context switch somewhere in FreeRTOS. I left it to the following investigation and next version. 

* HID mock device with one Interface literally is the HID Device with two Interfaces due to the configuration value CFG_TUD_HID. But Interfaces are absolutely the same and both contain Keyboard and Mouse, which could be used based on  ReportID.

If there are still some question feel free to ask them. I will be happy to explain more precisely.

roma-jam commented:
@igrr @Dazza0 
Feel free to review also. 
I would be very happy to get a feedback from you also.

Dazza0 commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1474 @@
+            hid_iface->ep_in = ep_in_desc->bEndpointAddress;
+            hid_iface->ep_in_mps = USB_EP_DESC_GET_MPS(ep_in_desc);
+        } else {
+            ESP_LOGE(TAG, "HID device EP IN %#X configuration error",
```
We're in a critical section here, so use `ESP_EARLY_LOGx` here instead. `ESP_LOGx` is a blocking API, so we'll usually get an assert you attempt to block in a critical section.

roma-jam replied:
Exactly! 
Thanks.

Dazza0 commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1474 @@
+    while (iface != NULL) {
+        tmp = STAILQ_NEXT(iface, tailq_entry);
+        if (iface->parent && (iface->parent->dev_addr == dev_addr)) {
+            ESP_ERROR_CHECK( hid_host_device_close(iface) );
+            ESP_ERROR_CHECK( hid_host_interface_shutdown(iface, false) );
```
Are we sure we want to free the interface and device automatically? As soon as we the interface object is freed, the `hid_host_device_handle_t` in the user's code is no longer valid (will point to an already freed object).

The safer way to handle this is to keep the interface object allocated (but have a state variable in the interface indicating that the interface is gone). The interface object remains allocated until the user explicitly calls `hid_host_device_close()`). Doing this will also save us from having to call `get_iface_by_handle()` at the start of every HID interface function, as we can assume that a provided handle will always have an allocated object behind it.

roma-jam replied:
Yeah....that the point :(
Now, there are two callback are available:
1. device connected (for every HID interface)
2. device event: input report, transfer error, device gone

If the user doesn't open a device after receiving device connected event, the callback with 'device gone' will not reach user. So that means, that there are two possible way of closing the device (interface in this case): 
1. If if was opened by user, close everything we can inside the driver, send event to user for delete object.
2. If it was not opened, then close everything we can inside the driver and delete object right now.

That is exactly the way as it implemented right now. 

But from some point, I am concerned about two callbacks and I want to make it simpler and merge them to one in the future. 
That would be a breaking change then, but I don't see the benefits from two callback right now (at least, anymore after several process of refactoring and eliminating other logic).

Dazza0 reviewed (commented):
@roma-jam `hid_host.h` API LGTM. Left some comments about thread safety and automatic device removal,

roma-jam commented on usb/usb_host_hid/README.md:
```diff
@@ -0,0 +1,48 @@
+[![Component Registry](https://components.espressif.com/components/espressif/usb_host_hid/badge.svg)](https://components.espressif.com/components/espressif/usb_host_hid)
+
+This directory contains an implementation of a USB HID Driver implemented on top of the [USB Host Library](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/api-reference/peripherals/usb_host.html).
+
```
Now the HID driver supports every HID compatible interface with bInterfaceClass = 0x03 (Human Interface Device). Any SubClass, any Protocol. 
But the keyboard event (key_pressed and key_released) is available only for Boot Protocol input report. 

I will add this information to the chapter ## Supported Devices. 
Thanks.

roma-jam commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -0,0 +1,275 @@
+ *
+ * @return dev_params           Pointer to a device parameters (USB port, Interface number, SubClass and Proto)
+ */
+hid_host_dev_params_t *hid_host_device_get_params(hid_host_device_handle_t hid_dev_handle);
```
Agreed. 
That was a long time ago when I added that function. After couple of examples I realized that it is better to do all reasonable-returnable function with esp_err_t return value.

roma-jam commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1474 @@
+
+static hid_driver_t *s_hid_driver;              /**< Internal pointer to HID driver */
+
+STAILQ_HEAD(devices, hid_host_device) hid_devices_tailq;
+STAILQ_HEAD(interfaces, hid_interface) hid_ifaces_tailq;
```
No reasons. Just a legacy from MSC example. 
Moved to the hid_driver_t. 

Thanks.

roma-jam commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1474 @@
+    // Delete HID compliant device
+    ESP_ERROR_CHECK( hid_host_uninstall_device(hid_device) );
+
+    return dev_addr;
```
Nope. 
Not anymore. 

Thanks.

roma-jam commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1474 @@
+ * @param[in] event    Client event message
+ * @param[in] arg      Argument, does not used
+ */
+static void client_event_cb(const usb_host_client_event_msg_t *event, void *arg)
+{
+    if (event->event == USB_HOST_CLIENT_EVENT_NEW_DEV) {
+        hid_host_device_init_attempt(event->new_dev.address);
+    } else if (event->event == USB_HOST_CLIENT_EVENT_DEV_GONE) {
+        hid_host_device_disconnected(event->dev_gone.dev_hdl);
+    }
+}
```
Yep, that is true. 
From that point (I didn't consider the calling path before) that is absolutely fair. 

Thanks for the tip.

Dazza0 commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -0,0 +1,1474 @@
+ */
+static void event_handler_task(void *arg)
+{
+    while (1) {
+        /* Here wee need a timeout 50 ms to handle end_client_event_handling flag
+         * during situation when all devices were removed and it is time to remove
+         * and destroy everything.
+         */
+        usb_host_client_handle_events(s_hid_driver->client_handle, pdMS_TO_TICKS(50));
+
+        if (s_hid_driver->end_client_event_handling) {
+            break;
+        }
+    }
+    usb_host_client_unblock(s_hid_driver->client_handle);
```
@roma-jam Doesn't `usb_host_client_unblock()` allow `usb_host_client_handle_events()` to return?

roma-jam replied:
Yes, it does.

Dazza0 reviewed (approved):
@roma-jam LGTM
